File: src/providers/openrouter.js
import { config } from "../config.js";

/**
 * Calls the OpenRouter API with the given messages and tools.
 * @param {Array} messages - The conversation history.
 * @param {Array} tools - The available tools.
 * @param {string} [model] - Optional model override.
 * @returns {Promise<Object>} The response message.
 */
export async function callOpenRouter(messages, tools, model = null) {
  const headers = {
    Authorization: `Bearer ${config.apiKey}`,
    "Content-Type": "application/json",
    "HTTP-Referer": "https://github.com/your-repo/agent", // Optional
    "X-Title": "Agent", // Optional
  };

  const body = {
    model: model || config.model,
    messages: messages,
    tools: tools,
    usage: { include: true },
    include_reasoning: true, // Request reasoning tokens
    provider: {
      allow_fallbacks: false,
    },
  };

  // Check for our special internal trigger for grounding
  const forceGrounding =
    tools &&
    tools.some(
      (t) => t.function && t.function.name === "__google_search_trigger__",
    );

  if (forceGrounding) {
    // Enable OpenRouter web search by appending :online to the model name
    // This is the documented way to enable web search for any model
    if (!body.model.endsWith(":online")) {
      body.model += ":online";
    }

    // Remove the dummy tool so the model doesn't try to call it
    // and instead uses the injected search results to generate content.
    if (body.tools) {
      body.tools = body.tools.filter(
        (t) => t.function.name !== "__google_search_trigger__",
      );
      // If no tools left, remove the tools property
      if (body.tools.length === 0) {
        delete body.tools;
      }
    }
  }

  const MAX_RETRIES = 3;
  let attempt = 0;

  while (attempt < MAX_RETRIES) {
    try {
      const response = await fetch(
        "https://openrouter.ai/api/v1/chat/completions",
        {
          method: "POST",
          headers: headers,
          body: JSON.stringify(body),
        },
      );

      if (!response.ok) {
        const errorText = await response.text();

        // Check for specific errors that might be worth retrying
        const isTransient = response.status >= 500 || response.status === 429;
        const isCorruptedThought =
          response.status === 400 &&
          errorText.includes("Corrupted thought signature");

        if (isTransient || isCorruptedThought) {
          console.warn(
            `Attempt ${attempt + 1} failed: ${
              response.status
            } - ${errorText}. Retrying...`,
          );
          attempt++;
          if (attempt < MAX_RETRIES) {
            await new Promise((resolve) =>
              setTimeout(resolve, 1000 * Math.pow(2, attempt)),
            ); // Exponential backoff
            continue;
          }
        }

        throw new Error(
          `OpenRouter API error: ${response.status} ${response.statusText} - ${errorText}`,
        );
      }

      const data = await response.json();

      if (data.usage) {
        console.log(
          "\x1b[2mToken Usage:",
          JSON.stringify(data.usage, null, 2),
          "\x1b[0m",
        );
      }

      const choice = data.choices[0];
      const message = choice.message;

      // Extract reasoning if present (OpenRouter puts it in 'reasoning' field)
      if (message.reasoning) {
        message.reasoning = message.reasoning;
      }

      return {
        message: message,
        usage: data.usage,
      };
    } catch (error) {
      // If it's a network error (fetch failed), retry
      if (error.name === "TypeError" && error.message === "fetch failed") {
        console.warn(
          `Attempt ${attempt + 1} failed: Network error. Retrying...`,
        );
        attempt++;
        if (attempt < MAX_RETRIES) {
          await new Promise((resolve) =>
            setTimeout(resolve, 1000 * Math.pow(2, attempt)),
          );
          continue;
        }
      }
      console.error("Failed to call OpenRouter:", error);
      throw error;
    }
  }
}

File: src/providers/openrouter.py
"""OpenRouter API provider."""

import requests
import json
import asyncio
import logging
import html
from prompt_toolkit import print_formatted_text
from prompt_toolkit.formatted_text import HTML
from ..config import config
from ..cache import apply_anthropic_cache

logger = logging.getLogger(__name__)


def prepare_messages_for_openrouter(messages: list, model: str) -> list:
    """Prepare messages for OpenRouter, including reasoning preservation for Gemini."""
    is_gemini = "gemini" in model.lower()
    prepared = []

    for msg in messages:
        new_msg = {"role": msg["role"]}

        # Handle content
        if "content" in msg:
            new_msg["content"] = msg["content"]

        # Handle tool calls (assistant messages)
        if "tool_calls" in msg:
            new_msg["tool_calls"] = msg["tool_calls"]

        # Handle tool results
        if "tool_call_id" in msg:
            new_msg["tool_call_id"] = msg["tool_call_id"]
        if "name" in msg:
            new_msg["name"] = msg["name"]

        # Preserve reasoning_details for Gemini models (required for tool calling)
        # This contains the encrypted thought signatures needed for multi-turn tool calls
        if is_gemini and "reasoning_details" in msg:
            new_msg["reasoning_details"] = msg["reasoning_details"]

        prepared.append(new_msg)

    return prepared


async def call_openrouter(messages: list, tools: list, model: str = None) -> dict:
    """Call OpenRouter API. Returns dict with 'message' and 'usage'."""
    effective_model = model or config.model

    # Detect provider type for optimizations
    is_anthropic = "anthropic" in effective_model or "claude" in effective_model
    is_openai = any(x in effective_model for x in ["openai", "gpt", "o1", "o3", "o4"])
    is_gemini = "gemini" in effective_model

    # Apply Anthropic caching if applicable
    if is_anthropic:
        apply_anthropic_cache(messages, effective_model)

    # Prepare messages (handles reasoning preservation for Gemini)
    prepared_messages = prepare_messages_for_openrouter(messages, effective_model)

    headers = {
        "Authorization": f"Bearer {config.openrouter_api_key}",
        "Content-Type": "application/json",
        "HTTP-Referer": "https://github.com/sonercirit/agent",
        "X-Title": "Agent",
    }

    body = {
        "model": effective_model,
        "messages": prepared_messages,
        "tools": tools if tools else None,
        "temperature": 0,
        "usage": {"include": True},
        "include_reasoning": True,
        "reasoning": {"effort": "high"},
        "provider": {"allow_fallbacks": False},
    }

    # Remove None values
    body = {k: v for k, v in body.items() if v is not None}

    # Lock to specific providers
    if is_anthropic:
        body["provider"] = {"order": ["Anthropic"], "allow_fallbacks": False}
    elif is_openai:
        body["provider"] = {"order": ["OpenAI"], "allow_fallbacks": False}
    elif is_gemini:
        body["provider"] = {"order": ["Google AI Studio"], "allow_fallbacks": False}

    # Handle Google Search trigger
    if tools and any(t["function"]["name"] == "__google_search_trigger__" for t in tools):
        if not body["model"].endswith(":online"):
            body["model"] += ":online"
        body["tools"] = [t for t in tools if t["function"]["name"] != "__google_search_trigger__"]
        if not body["tools"]:
            body.pop("tools", None)

    # Retry loop
    for attempt in range(3):
        try:
            response = await asyncio.to_thread(
                requests.post, "https://openrouter.ai/api/v1/chat/completions", headers=headers, json=body, timeout=120
            )

            if response.status_code != 200:
                if response.status_code >= 500 or response.status_code == 429:
                    logger.warning(f"Attempt {attempt + 1} failed: {response.status_code}. Retrying...")
                    await asyncio.sleep(2**attempt)
                    continue
                raise Exception(f"OpenRouter API error: {response.status_code} - {response.text}")

            data = response.json()

            if "error" in data:
                logger.warning(f"API error: {data['error']}. Retrying...")
                await asyncio.sleep(2**attempt)
                continue

            if "choices" not in data or not data["choices"]:
                logger.warning("No choices in response. Retrying...")
                await asyncio.sleep(2**attempt)
                continue

            usage = data.get("usage", {})
            if usage:
                print_formatted_text(HTML(f"<style fg='#666666'>Tokens: {html.escape(json.dumps(usage))}</style>"))

            msg = data["choices"][0]["message"]

            # Log reasoning details preservation for debugging
            if is_gemini and msg.get("reasoning_details"):
                logger.debug(f"Response has reasoning_details: {len(msg['reasoning_details'])} items")

            return {"message": msg, "usage": usage}

        except requests.exceptions.RequestException as e:
            logger.warning(f"Attempt {attempt + 1} failed: {e}")
            await asyncio.sleep(2**attempt)

    raise Exception("Failed to call OpenRouter API after 3 retries.")
